### PR TITLE
Strip leading directory path from request

### PR DESCRIPTION
**Pull request by elJosho**
This will strip the leading directory from a request if configured.  It's useful when you're trying to expose a bucket at a specific path on an ALB.  For example, if I have an ALB rule that forwards all traffic from "www.mysite.com/mybucket" to the S3 gateway, and I want to request a file in the root of my bucket, "myfile.txt".  Making a request to "www.mysite.com/mybucket/myfile.txt" will fail as the S3 gateway will look for the file inside a folder named "mybucket", rather than at the root of the bucket.  

After this change, setting the new env var to
`STRIP_LEADING_DIRECTORY_PATH = /mybucket`
will send all requests to the root of the bucket.

Hoping this helps someone, it worked well for me.  Please be gentle if I did anything especially stupid, this is my first time submitting a PR to a public repo.

### CODE REVIEW

**Comment by dekobon** [comment]
@elJosho Thank you for the contribution! I'm going to review it soon.

**Review by 4141done — COMMENTED** [review]
Thank you for your contribution and I apologize for the delay - I'm a new maintainer and it took me a little more time to look over your suggested change. Also congratulations on making your first public contribution!

In general I think this looks like a good use case, and we actually have an existing parameter that does this for the directory listing function: `DIRECTORY_LISTING_PATH_PREFIX`.

I think the two configuration options would make a good pair.
`STRIP_LEADING_DIRECTORY_PATH` -> For retrieving individual files
`DIRECTORY_LISTING_PATH_PREFIX` -> For listing directories

Before we can merge there are a few things I'd like to add:
1. I think you need a `default` clause in your new `map` statement. - otherwise all requests without the prefix will fail if you have a prefix set. (see comment)
2. Can you think of a name that you that might more closely match `DIRECTORY_LISTING_PATH_PREFIX` that accurately describes your feature?  That way others can recognize them as related. 
3. Can you add some usage instructions next to those for `DIRECTORY_LISTING_PATH_PREFIX` ([link](https://github.com/nginxinc/nginx-s3-gateway/blob/master/docs/getting_started.md#prefixing-list-results))

I'd also like to add some integration tests but I'm still wrapping my head around the setup.  I'm looking in to learning the ropes by writing some against your change. I'll have news on that as we work through the above items.

**Comment by 4141done** [comment]
I worked through how to add your configuration option to the integration test suite.  You can view the patch here:
https://gist.github.com/4141done/49c218ef8b5fec97060c859d45a65ac5

You can either add it to your branch (I suggest first rebasing on the upstream `master` branch as some changes have gone in since then) or with your permission I can push those changes to your fork.

When running those integration tests I have confirmed the necessity of a `default` clause to the added map statement but it passes all tests after that! 🎉

**Comment by elJosho** [comment]
Thanks so much for looking at this!  I've had a bit of a week so haven't had the chance to address your comments but will as soon as I can.

**Comment by 4141done** [comment]
> Thanks so much for looking at this! I've had a bit of a week so haven't had the chance to address your comments but will as soon as I can.

Sounds good!  We appreciate you taking the time to contribute. We can work together to get this merged on a timeline that works for you 🐳

File: common/etc/nginx/nginx.conf [hunk 1]
```diff
@@ -31,6 +31,7 @@ env PROXY_CACHE_VALID_NOTFOUND;
 env PROXY_CACHE_VALID_FORBIDDEN;
 env HEADER_PREFIXES_TO_STRIP;
 env FOUR_O_FOUR_ON_EMPTY_BUCKET;
+env STRIP_LEADING_DIRECTORY_PATH;
```
Comment (liyahui2012) [inline common/etc/nginx/nginx.conf:34]: Suggestion: 
The variable "STRIP_LEADING_DIRECTORY_PATH" can be set as a toggle switch, and the PATH can be directly use "S3_BUCKET_NAME".

**Comment by elJosho** [comment]
Hi, sorry again for the delay on this, health problems.  I rebased my branch off master and attempted to apply your patch, however it failed on patching test.sh, likely b/c I've waited too long.  Would you be able to apply it to my fork as you mentioned?  Also, any suggestions on a different name for the property?  I'm terrible at naming stuff, it look me forever to come up with STRIP_LEADING_DIRECTORY_PATH but would happily rename to whatever you think makes sense.

Thanks!

**Comment by 4141done** [comment]
Hi @elJosho I'm sorry to hear about your health and hope you are well. It looks like the updates against master made things a little confusing so I pulled it out into another branch here https://github.com/nginxinc/nginx-s3-gateway/pull/185

Would you mind testing it to make sure that it works for your use case?

Upon regaining my context on this feature, I think the name of the configuration variable actually sounds fine.  Once we're sure it works for you then I think we can just merge the other branch with some small docs additions.  Your authorship will be maintained since I cherry picked the commits over from your fork.

**Comment by elJosho** [comment]
Pulled down your branch and it works for me.  Thanks so much again for all your help with this!

**Comment by 4141done** [comment]
Closing in favor of https://github.com/nginxinc/nginx-s3-gateway/pull/185 which is the same but resolved some merge issues.

**Comment by 4141done** [comment]
@elJosho Thank you so much for your effort with this pull request and the considered feature proposal. It has been merged to mainline and new images containing the change will be available shortly.

**Comment by elJosho** [comment]
@4141done thank you for your help with this, I really appreciate it!

**Pull request closed without merging**